services/web/server/src/simcore_service_webserver/projects/projects_nodes_utils.py

```diff
@@ -0,0 +1,76 @@
+import logging
+from typing import Optional
+
+from aiohttp import web
+from models_library.errors import ErrorDict
+from pydantic.types import PositiveInt
+from servicelib.logging_utils import log_decorator
+from servicelib.utils import logged_gather
+
+from .projects_utils import project_get_depending_nodes
+
+log = logging.getLogger(__name__)
+
+
+@log_decorator(logger=log)
+async def update_node_outputs(
+    app: web.Application,
+    user_id: PositiveInt,
+    project_uuid: str,
+    node_uuid: str,
+    outputs: dict,
+    run_hash: Optional[str],
+    node_errors: Optional[list[ErrorDict]],
+    *,
+    ui_changed_keys: Optional[set[str]],
+) -> None:
+    # unsure how to deal with this circular dependency
+    # this function is called from:
+    # - `projects/projects_db.py`
+    # - `computation_comp_tasks_listening_task.py` (was originally here)
+    from . import projects_api
```

Comment:
This is really not good. We should have a look at it together. The more we delay solving this the worse it becomes
Please add an issue to follow up on this and make sure there is a FIXME here

Reply:
Added FIXME and created an issue
https://github.com/ITISFoundation/osparc-simcore/issues/3069

Reply:
I checked and reduced a bit the dependency but there is still a loop that IMO reveals a design problem in the changes of your PR

![image](https://user-images.githubusercontent.com/32402063/170026009-de8531ae-ff95-44da-aaf0-fe07d5f29ab5.png)

The problem i see is that you are including plugin-level API (i.e. ``projects_api``) in the implementation of db-API repository (i.e. ``projects_db``) while it should be the other way around i.e. the plugin-level API **uses** the db-API. 

More details on the code follow ...

Reply:
Thanks for this. How did you mange to find the loop, just by visually checking?

Reply:
RTFM? :-D ... jokes aside, I did it both visually and following the code. 

But there is also the option to find them automatically [--show-cycles](https://github.com/thebjorn/pydeps#import-cycles)

Reply:
NOTE: Perhaps add an example of ``--show-cycles`` in the doc of the script?